src/tls/keypair.h

```diff
@@ -17,67 +17,159 @@
+  static constexpr bool prefer_bitcoin_secp256k1 = true;
 
-#  define HASH(data_ptr, data_size, hash_ptr) \
-    do \
-    { \
-      mbedtls_sha512_ret(data_ptr, data_size, hash_ptr, false); \
-    } while (0)
+  // Helper to access elliptic curve id from context
+  inline mbedtls_ecp_group_id get_ec_from_context(const mbedtls_pk_context& ctx)
+  {
+    return mbedtls_pk_ec(ctx)->grp.id;
+  }
 
-#elif CURVE_CHOICE_SECP256K1_MBEDTLS || CURVE_CHOICE_SECP256K1_BITCOIN
+  // Get mbedtls elliptic curve for given CCF curve implementation
+  inline mbedtls_ecp_group_id get_ec_for_curve_impl(CurveImpl curve)
+  {
+    switch (curve)
+    {
+      case CurveImpl::secp384r1:
+      {
+        return MBEDTLS_ECP_DP_SECP384R1;
+      }
+      case CurveImpl::curve25519:
+      {
+        return MBEDTLS_ECP_DP_CURVE25519;
+      }
+      case CurveImpl::secp256k1_mbedtls:
+      case CurveImpl::secp256k1_bitcoin:
+      {
+        return MBEDTLS_ECP_DP_SECP256K1;
+      }
+      default:
+      {
+        throw std::logic_error("Unhandled curve type");
+      }
+    }
+  }
 
-#  define HASH(data_ptr, data_size, hash_ptr) \
-    do \
-    { \
-      mbedtls_sha256_ret(data_ptr, data_size, hash_ptr, false); \
-    } while (0)
+  // Get message digest algorithm to use for given elliptic curve
+  inline mbedtls_md_type_t get_md_for_ec(mbedtls_ecp_group_id ec)
+  {
+    switch (ec)
+    {
+      case MBEDTLS_ECP_DP_SECP384R1:
+      {
+        return MBEDTLS_MD_SHA384;
+      }
+      case MBEDTLS_ECP_DP_CURVE25519:
+      {
+        return MBEDTLS_MD_SHA512;
+      }
+      case MBEDTLS_ECP_DP_SECP256K1:
+      {
+        return MBEDTLS_MD_SHA256;
+      }
+      default:
+      {
+        throw std::logic_error("Unhandled curve type");
+      }
+    }
+  }
 
-#endif
+  /**
+   * Hash the given data, with an algorithm chosen by key type
+   *
+   * @return 0 on success
+   */
+  inline int do_hash(
+    const mbedtls_pk_context& ctx,
+    const uint8_t* data_ptr,
+    size_t data_size,
+    HashBytes& o_hash)
+  {
+    const auto ec = get_ec_from_context(ctx);
 
-namespace tls
-{
-#if CURVE_CHOICE_SECP384R1
-  static constexpr mbedtls_md_type_t MD_TYPE = MBEDTLS_MD_SHA384;
-  static constexpr size_t MD_SIZE = 384 / 8;
-#elif CURVE_CHOICE_CURVE25519
-  static constexpr mbedtls_md_type_t MD_TYPE = MBEDTLS_MD_SHA512;
-  static constexpr size_t MD_SIZE = 512 / 8;
-#elif CURVE_CHOICE_SECP256K1_MBEDTLS || CURVE_CHOICE_SECP256K1_BITCOIN
-  static constexpr mbedtls_md_type_t MD_TYPE = MBEDTLS_MD_SHA256;
-  static constexpr size_t MD_SIZE = 256 / 8;
-#endif
+    switch (ec)
+    {
+      case MBEDTLS_ECP_DP_SECP384R1:
+      {
+        constexpr auto hash_size = 384 / 8;
+        if (o_hash.size() < hash_size)
+          o_hash.resize(hash_size);
 
-  static constexpr size_t REC_ID_IDX = 64;
+        return mbedtls_sha512_ret(data_ptr, data_size, o_hash.data(), true);
+      }
+      case MBEDTLS_ECP_DP_CURVE25519:
+      {
+        constexpr auto hash_size = 512 / 8;
```

Comment:
Not really part of this refactoring but I think we could retrieve this from `mbedtls` directly, e.g. https://github.com/ARMmbed/mbedtls/blob/8266acacc8d6e1c65fba9a048f56339d0827b2fe/library/md_wrap.c#L572 ?

Reply:
Good spot. Turns out remove this entire switch and let mbedtls find the sizes and algorithm for us.